book/outline.txt

```diff
@@ -0,0 +1,465 @@
+# -*- mode: python -*-
+
+# This file is a template for the code outlines in the book;
+# definitions will appear in the outline in the same order that they appear here
+
+# In general, the format is interpreted as follows:
+#  - Whitespace lines and comments don't matter
+#  - Classes, functions, and constants are matched by name with the relevant prefix
+#  - So function arguments don't matter, but you can have them if you want
+#  - A line with two constants on it, like `WIDTH, HEIGHT`, will print with both constants on one line, in that order
+#  - If a definition is generated but not in this template you'll get an error
+
+# Networking
+COOKIE_JAR
+class URL:
+    def __init__
+    def request
+    def resolve
+    def origin
+    def __str__
+
+# HTML tree
+class Text:
+    def __init__
+    def __repr__
+class Tag:
+    def __init__
+class Element:
+    def __init__
+    def __repr__
+def print_tree
+def tree_to_list
+def is_focusable(node)
+def get_tabindex(node)
+
+# HTML parser
+def show(body)
+def lex(body)
+class HTMLParser:
+    SELF_CLOSING_TAGS
+    HEAD_TAGS
+    def __init__(body)
+    def parse()
```

Comment:
How did you choose the order of the methods? Something better than alphabetical?

Reply:
At the top level, sort of attempting to follow the browser rendering pipeline back to front, with similar methods & classes together. Within a class there were a lot more judgement calls but vaguely similar things should be together, and, say, layout should go before paint.